soc/arm/ti_simplelink/cc13x2_cc26x2/power.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2019 Linaro Limited.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <init.h>
+#include <power/power.h>
+
+#include <ti/drivers/Power.h>
+#include <ti/drivers/power/PowerCC26X2.h>
+
+#include <ti/drivers/dpl/ClockP.h>
+
+#include <ti/devices/cc13x2_cc26x2/driverlib/cpu.h>
+#include <ti/devices/cc13x2_cc26x2/driverlib/vims.h>
+#include <ti/devices/cc13x2_cc26x2/driverlib/sys_ctrl.h>
+
+#include <logging/log.h>
+#define LOG_LEVEL CONFIG_SOC_LOG_LEVEL
+LOG_MODULE_REGISTER(soc);
+
+/* Configuring TI Power module to not use its policy function (we use Zephyr's
+ * instead), and disable oscillator calibration functionality for now.
+ */
+const PowerCC26X2_Config PowerCC26X2_config = {
+	.policyInitFxn      = NULL,
+	.policyFxn          = NULL,
+	.calibrateFxn       = NULL,
+	.enablePolicy       = false,
+	.calibrateRCOSC_LF  = false,
+	.calibrateRCOSC_HF  = false
+};
+
+extern PowerCC26X2_ModuleState PowerCC26X2_module;
+
+/*
+ * Power state mapping:
+ * SYS_POWER_STATE_SLEEP_1: Idle
+ * SYS_POWER_STATE_SLEEP_2: Standby
+ */
+
+/* Invoke Low Power/System Off specific Tasks */
+void sys_set_power_state(enum power_states state)
+{
+#ifdef CONFIG_SYS_POWER_SLEEP_STATES
+#ifdef CONFIG_HAS_SYS_POWER_STATE_SLEEP_1
+	u32_t modeVIMS;
+	u32_t constraints;
+#endif
+#endif
+
+	LOG_DBG("SoC entering power state %d", state);
+
+	/* Switch to using PRIMASK instead of BASEPRI register, since
+	 * we are only able to wake up from standby while using PRIMASK.
+	 */
+	/* Set PRIMASK */
+	CPUcpsid();
+	/* Set BASEPRI to 0 */
+	irq_unlock(0);
+
+	switch (state) {
+#ifdef CONFIG_SYS_POWER_SLEEP_STATES
+#ifdef CONFIG_HAS_SYS_POWER_STATE_SLEEP_1
+	case SYS_POWER_STATE_SLEEP_1:
+		/* query the declared constraints */
+		constraints = Power_getConstraintMask();
+		/* 1. Get the current VIMS mode */
+		do {
+			modeVIMS = VIMSModeGet(VIMS_BASE);
+		} while (modeVIMS == VIMS_MODE_CHANGING);
+
+		/* 2. Configure flash to remain on in IDLE or not and keep
+		 *    VIMS powered on if it is configured as GPRAM
+		 * 3. Always keep cache retention ON in IDLE
+		 * 4. Turn off the CPU power domain
+		 * 5. Ensure any possible outstanding AON writes complete
+		 * 6. Enter IDLE
+		 */
+		if ((constraints & (1 << PowerCC26XX_NEED_FLASH_IN_IDLE)) ||
+			(modeVIMS == VIMS_MODE_DISABLED)) {
+			SysCtrlIdle(VIMS_ON_BUS_ON_MODE);
+		} else {
+			SysCtrlIdle(VIMS_ON_CPU_ON_MODE);
```

Comment:
```
void SysCtrlIdle(uint32_t vimsPdMode)
{
    // Configure the VIMS mode
    HWREG(PRCM_BASE + PRCM_O_PDCTL1VIMS) = vimsPdMode;

    // Always keep cache retention ON in IDLE
    PRCMCacheRetentionEnable();

    // Turn off the CPU power domain, will take effect when PRCMDeepSleep() executes
    PRCMPowerDomainOff(PRCM_DOMAIN_CPU);

    // Ensure any possible outstanding AON writes complete
    SysCtrlAonSync();

    // Invoke deep sleep to go to IDLE
    PRCMDeepSleep();
}
```
`Turn off the CPU power domain` means cpu power gating? @vanti

Reply:
Yes. The device is partitioned into multiple power domains, and this code specifically gating off the CPU power domain, in addition to clock gating.

Reply:
so that it will lose cpu content, hardware will help store/restore cpu content?

Reply:
Yes the CPU module has retention which allows its state to be preserved.

Reply:
ok, it seems NVIC isn't in the same power domain with CPU because in Power_sleep you have to save NVIC registers.

Reply:
@vanti sorry for misreading the code. Now I understand this state, and I'm fine with defining it sleep state.